ecommerce/taxes/lib/taxes/vat_rate_catalog.rb

```diff
@@ -16,13 +16,17 @@ def vat_rate_for(product_id)
     end
 
     def vat_rate_by_code(vat_rate_code)
```

Comment:
Outside of scope of this task, but this method got me thinking.

This method is called in the [AddAvailableVatRateHandler ](https://github.com/RailsEventStore/ecommerce/blob/master/ecommerce/taxes/lib/taxes/services.rb#L50)

It is called to check if tax rate with given code already exists. If not, then the handler publishes an event.

First of all, it breaks the tell don't ask rule. In current implementation there is possibility that two `AvailableVatRateAdded` events will be published. Therefore, the system will be in unexpected state.

The question is if that is a major or minor problem.

Since this if statement has been added for some reason, let's assume it would a major issue if the vat rate code is duplicated.

To reproduce this behaviour you can follow those steps:

```ruby

rails-application(dev)* fiber = Fiber.new do
rails-application(dev)*   unless event_store.read.stream("AvailableVatRate$#{vat_rate_code}").last
rails-application(dev)*     Fiber.yield
rails-application(dev)*     event_store.publish(AvailableVatRateAdded.new(data: { vat_rate_code: }), stream_name: "AvailableVatRate$#{vat_rate_code}")
rails-application(dev)*   end
rails-application(dev)> end
rails-application(dev)> fiber.resume
  RubyEventStore::ActiveRecord::EventInStream Load (0.6ms)  SELECT "event_store_events_in_streams".* FROM "event_store_events_in_streams" WHERE "event_store_events_in_streams"."stream" = $1 ORDER BY "event_store_events_in_streams"."id" DESC LIMIT $2  [["stream", "AvailableVatRate$60"], ["LIMIT", 1]]
=> nil # <========= The if statement was checked here, there was no event in stream
rails-application(dev)> event_store.publish(AvailableVatRateAdded.new(data: { vat_rate_code: }), stream_name: "AvailableVatRate$#{vat_rate_code}")
rails-application(dev)> event_store.read.stream("AvailableVatRate$#{vat_rate_code}").to_a.count
=> 2
```
How do you think we could deal with that? (tip: look at the core concepts section in RES documentation)

Reply:
> Is this what was concerning you?
Yes, exactly. Good job :)

I think that code is in ok shape now. I wouldn't optimize it anymore at this stage.

There are no business rules or invariants that would justify introducing an aggregate here. I would leave it as it is and observe how this code changes over time. 

If, say there is another place that needs the piece of code that: reads the vat rate stream, checks if last vat rate event is X and does something, then I'd think about introducing some concept (not necessarily an aggregate) to keep it together. Perhaps Domain Service?